File: Evol'artist/src/simulEngine/Creature.cpp
//
//  Creature.cpp
//  Evol'artist
//
//  Created by Emma Barme on 28/09/2017.
//  Copyright GNU General Public License v3.0
//
//  Last modified by Emma Barme on 29/09/2017

#include "Creature.hpp"
#include "World.hpp"

Creature::Creature(const int id, const std::string genome, const int x, const int y, std::map<std::string, int> env): m_id(id), m_genome(genome) {
    //Initialise attributes by default
    
    m_genes = interpretGenome();
    
    // Get fitness
    FitnessFunctions f = FitnessFunctions();
    m_fitness = f.getFitness(m_genes, env);
    
    //Coordinates
    m_x = x;
    m_y = y;
}

Creature::Creature(const Creature& creature): m_id(creature.m_id), m_x(creature.m_x), m_y(creature.m_y), m_genome(creature.m_genome), m_genes(creature.m_genes), m_fitness(creature.m_fitness) {}

Creature Creature::operator=(const Creature& creature) {
    return Creature(creature);
}

const int Creature::getId() const {
    return m_id;
}

const std::string Creature::getGenome() const {
    return m_genome;
}

const int Creature::getFitness() const {
    return m_fitness;
}

const std::pair<int, int> Creature::getCoord() const {
    return std::pair<int, int> (m_x, m_y);
}

std::vector<std::string> Creature::interpretGenome() {
    // Get list of genes using regular expressions
    std::regex genesRegex("A[BCD]{2,3}E");
    auto genesBegin = std::sregex_iterator(m_genome.begin(), m_genome.end(), genesRegex);
    auto genesEnd = std::sregex_iterator();
    std::vector<std::string> genes;
    for (std::sregex_iterator i = genesBegin; i != genesEnd; ++i) {
        std::smatch match = *i;
        std::string gene = match.str();
        gene = gene.substr(1,gene.size() - 2); //Strip A and E
        genes.push_back(gene);
    }
    
    return genes;
}

const std::pair<int, int> Creature::move(const int worldSize) {
    return std::make_pair(m_x, m_y);
}

File: UnityProject/Assets/Scripts/Simulation/Creature.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
namespace Application
{
    public class Creature
    {
        /*
         * Identifier of the creature
         */
        private int m_id;

        /*
         * Single chromosome genome
         */
        private string m_genome;

        /*
         * List of genes by alphabetical order
         */
        private List<string> m_genes;

        /*
         * Fitness (between 0 and 100)
         */
        private int m_fitness;

        /*
         * Horizontal coordinate of the creature in the world grid
         */
        private int m_x;

        /*
         * Vertical coordinate of the creature in the world grid
         */
        private int m_y;

        /*
         * Creature constructor with specified coordinates
         * @param id Creature id
         * @param genome Single chromosome genome
         * @param x Horizontal coordinate
         * @param y Vertical coordinate
         * @param env Dict describing environment
         */
        public Creature(int id, string genome, int x, int y, Dictionary<string, int> env)
        {
            //Initialise attributes
            m_id = id;
            m_genome = genome;
            Test.Print("pre-interpret");
            m_genes = interpretGenome();

            // Get fitness
            FitnessFunctions f = new FitnessFunctions();
            Test.Print("between-fitness");
            m_fitness = f.getFitness(m_genes, env);
            Test.Print("post-fitness");

            //Coordinates
            m_x = x;
            m_y = y;
        }

        /*
         * Creature copy constructor
         * @param creature The Creature object to copy
         */
        public Creature(Creature creature)
        {
            //Initialise attributes
            m_id = creature.m_id;
            m_genome = creature.m_genome;
            m_genes = creature.m_genes;
            m_fitness = creature.m_fitness;
            //Coordinates
            m_x = creature.m_x;
            m_y = creature.m_y;
        }

        /*
         * Identifier accessor
         * @return m_id
         */
        public int getId() {
            return m_id;
        }

        /*
         * Genome accessor
         * @return m_genome
         */
        public string getGenome() {
            return m_genome;
        }

        /*
         * Fitness accessor
         * @return m_fitness
         */
        public int getFitness() {
            return m_fitness;
        }

        /*
         * Coordinates accessor
         * @return The pair of horizontal and vertical coordinates
         */
        public Tuple<int, int> getCoord() {
            return new Tuple<int, int> (m_x, m_y);
        }

        /*
         * Interpret genome as genes
         * @return A list of valide gene strings
         */
        public List<string> interpretGenome() {
            // Get list of genes using regular expressions
            Regex genesRegex = new Regex("A[BCD]{2,3}E");
            List<string> genes = new List<string>();
            Test.Print(m_genome);
            Match match = genesRegex.Match(m_genome);
            while (match.Success) {
                Test.Print(match.Value);
                string gene = match.Value;
                gene = gene.Substring(1, gene.Length - 2);
                genes.Add(gene);
                match = match.NextMatch();
            }
            Test.Print("End while");
            return genes;
        }

        /*
         * Check if has a specific gene
         * @param gene The string of the gene
         * @return True if the gene is present
         */
        public bool hasGene(string gene) {
            return m_genes.Exists(s => s == gene);
        }
    }
}
